tests/integration/application-charm/src/charm.py

```diff
@@ -0,0 +1,162 @@
+
+
+class ApplicationCharm(CharmBase):
+    """Application charm that relates to MySQL Router charm."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.framework.observe(self.on.start, self._on_start)
+
+        self.database_name = f"{self.app.name.replace('-', '_')}_test_database"
+        self.database_requires = DatabaseRequires(self, REMOTE, self.database_name)
+        self.framework.observe(
+            self.database_requires.on.database_created, self._on_database_created
+        )
+        self.framework.observe(self.on.get_inserted_data_action, self._get_inserted_data)
+
+    # =======================
+    #  Handlers
+    # =======================
+
+    def _on_start(self, _) -> None:
+        """Handle the start event by setting the charm in waiting status.
+
+        If the mysqlrouter application is scaling up, then confirm that this
+        application can connect to the database from the bootstrapped mysqlrouter.
+        """
+        self.unit.status = WaitingStatus("Waiting for mysqlrouter relation")
+
+        peer_databag = self._peers.data[self.app]
+        database_config = peer_databag.get("database_read_config")
+        inserted_value = peer_databag.get("inserted_value")
+
+        if database_config and inserted_value:
+            database_config = json.loads(database_config)
+
+            with MySQLConnector(database_config) as cursor:
+                output = self._query_data(
+                    cursor,
+                    f"SELECT data FROM {self.database_name}.app_data WHERE data = '{inserted_value}'",
+                )
+
+            if output[0] == inserted_value:
+                self.unit.status = ActiveStatus()
+            else:
+                self.unit.status = BlockedStatus("Failed to read data")
+
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Handle the database created event."""
+        if not self.unit.is_leader():
+            return
+
+        logger.info("Received the database created event")
+
+        config = {
+            "user": event.username,
+            "password": event.password,
+            "host": event.endpoints.split(":")[0],
+            "port": event.endpoints.split(":")[1],
+            "database": self.database_name,
+            "raise_on_warnings": False,
+        }
+
+        with MySQLConnector(config) as cursor:
+            self._create_test_table(cursor)
+
+            random_value = generate_random_chars(255)
+            self._insert_test_data(cursor, random_value)
+
+        read_only_config = copy.deepcopy(config)
+        read_only_config["host"] = event.read_only_endpoints.split(":")[0]
+        read_only_config["port"] = event.read_only_endpoints.split(":")[1]
+
+        with MySQLConnector(read_only_config) as cursor:
+            output = self._query_data(
+                cursor,
+                f"SELECT data FROM {self.database_name}.app_data WHERE data = '{random_value}'",
```

Comment:
Optional: a test to ensure that `read_only_endpoints` field is really a read only endpoint (try to write some data and catch it with `pytest.raises` - I agree that is not so beautiful in a charm code and it's better to have such assertion on the test code, but probably it's ok in a tester charm).

Reply:
i added a check in the application charm itself in b22efe12e0d56b70d439b490de176040b0e1bd8f. if reads can occur from the read-only endpoint, the charm moves into blocked status, and the test then fails